Add context to object save in JS SDK

mtrezza opened the pull request:
Added description for new feature to expose Cloud Code trigger context in `Parse.Object.save()`.

---
Don't merge until

- [x] Change in JS SDK https://github.com/parse-community/Parse-SDK-JS/pull/1150
- [x] Change in Parse Server https://github.com/parse-community/parse-server/pull/6626
- [x] Replace required parse server version https://github.com/parse-community/docs/pull/730/commits/453d2fd5128c183d183b7ceaf4170b8b0833f5fb

mtrezza committed:
Merge commit 'e129790ed587819ecfcaed59fc7ff90ec6ffcfa1' into gh-pages

* commit 'e129790ed587819ecfcaed59fc7ff90ec6ffcfa1': (84 commits)
  Add `cascadeSave` option in ParseObject to JS docs (#664)
  Update apple sign in documentation (#662)
  Bump mixin-deep from 1.3.1 to 1.3.2 (#670)
  Update GraphQL docs with latest changes (#669)
  remove a bunch of parse cloud references (#654)
  docs: fixed examples for pipeline in aggregation (#661)
  Add beta notice to the GraphQL Guide (#659)
  iOS Documentation Revamp (#650)
  Fix doc for creating a subscription in javascript (#658)
  GraphQL Customisation (#652)
  added PhantAuth authData docs (#655)
  Fix typo in keys.md (#656)
  Fix typo in cloud-code.md (#657)
  add reference to afterFind, add more detail on predefined classes (#653)
  Add `keys` and `excludeKeys` options to REST docs (#649)
  document push type and priority keys for notifications (#639)
  Documentation for Sign In With Apple (#648)
  Documentation for isDataAvailable (#647)
  Remove line for 3.x (#646)
  Add documentation for query.withCount (#645)
  ...

mtrezza committed:
Merge remote-tracking branch 'upstream/gh-pages' into gh-pages

mtrezza committed:
Merge commit 'cc489df4a3d8544a52148c079ecaee3d1762108d' into gh-pages

* commit 'cc489df4a3d8544a52148c079ecaee3d1762108d':
  Adding microsoft oauth doc (#698)
  Encrypting Current User and Local Storage (#695)
  Update live-query.md (#696)
  Update config.md (#689)
  Fix User Subclass Documentation (#690)
  remove info about parse.com compatibility (#686)

mtrezza committed:
Merge remote-tracking branch 'upstream/gh-pages' into gh-pages

mtrezza committed:
added context to object save in JS SDK

TomWFox commented on _includes/js/objects.md:
```diff
@@ -199,6 +199,29 @@ teamMember.save(null, { cascadeSave: false });
 
+### Cloud Code context
+
+You may pass a `context` that is accessible in Cloud Code `beforeSave` and `afterSave` triggers for that `Parse.Object`. This is useful if you want to condition certain operations in Cloud Code triggers on ephemeral information that should not be saved with the `Parse.Object` in the database. The context is ephemeral in the sense that it vanishes after the Cloud Code triggers for that particular  ...
```
```suggestion
You may pass a `context` dictionary that is accessible in Cloud Code `beforeSave` and `afterSave` triggers for that `Parse.Object`. This is useful if you want to condition certain operations in Cloud Code triggers on ephemeral information that should not be saved with the `Parse.Object` in the database. The context is ephemeral in the sense that it vanishes after the Cloud Code triggers for that particular `Parse.Object` have executed. For example:
```
Or ‘`context` object’ if you prefer?

mtrezza replied:
The context is currently in `afterSave Triggers` > `Using Request Context`.

Should I move it into its own main section, since it belongs to both triggers and beyond?

TomWFox replied:
That sounds like a plan 👌

mtrezza replied:
Check it out, looks much cleaner now.

- added headings per trigger group
- moved context of save triggers into own section

![image](https://user-images.githubusercontent.com/5673677/79671026-661ef000-81c7-11ea-8ecf-02ff9a11e380.png)

TomWFox replied:
That’s a lot better!

TomWFox reviewed (changes requested):
Looks good, one wee suggestion.

Also, I think it would be worth including this in the cloud code guide too. You could link to the content in the JS guide or vice versa but considering it’s a relatively small section I would just copy and paste it in.

mtrezza committed:
Update _includes/js/objects.md

Co-Authored-By: Tom Fox <13188249+TomWFox@users.noreply.github.com>

mtrezza committed:
added context info to cloud code

mtrezza committed:
restructuring cloud code doc

- added headings per trigger group
- moved context of save triggers into own section

TomWFox reviewed (dismissed):
Barring any changes to the implementation, this looks fab.

I love when people organise all their PRs nicely & take care of the docs early in the process, it makes it sooo much easier for us!

mtrezza commented:
The first thing I think of before starting a PR is "Which changes in the docs would make Tom happy?" 😌

TomWFox commented:
❤️ 

I think I might actually make a bot that comments that on every PR to a Parse repo 😅😝

mtrezza commented:
@TomWFox Do we have a guideline for adding the required minimum parse server version for a new feature to the docs?

I rarely see something like “Requires Parse Server v1.2”. Should that added or do we expect users to read the change log?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

TomWFox commented:
Good point, we don’t have a guideline for this but probably should.

When there is a clear section for a new feature in the docs I think it makes sense to add a note, like in some sections of the Cloud Code guide...
![IMG_0901](https://user-images.githubusercontent.com/13188249/79687123-fa776a00-823c-11ea-861f-bdd1ce2fa67e.jpeg)

I’ve been bitten by this when I was trying to use request context only to realise that it wasn’t included in the version I was running.

mtrezza commented:
Yes, I also found some of these:

![image](https://user-images.githubusercontent.com/5673677/79690871-1ab72100-825d-11ea-88a3-ea0ea4cef7f6.png)

mtrezza committed:
Added required parse server version (tbd)

- Insert version before merging

TomWFox committed:
add parse server version

mtrezza commented:
Uh, I missed track of that one, thanks @TomWFox

TomWFox commented:
No probs, me too!
